# Setup

In [1]:
import os
import re
import gzip

import numpy as np
import pandas as pd
from scipy.stats import fisher_exact
from statsmodels.stats.multitest import multipletests

from Bio import SeqIO

from IPython.display import display

DIR = r'c://downloads'

# Q1

According to [Wikipedia](https://en.wikipedia.org/wiki/PDZ_domain), PDZ domains play a key role in anchoring receptor proteins in the membrane to cytoskeletal components and in the formation and function of signal transduction complexes, including transport and ion channel signaling.

# Q2

To find keywords significantly associated with the PDZ domain, we will carry out the analysis at the resolution of proteins (not domains) since proteins are the entities that are annotated with keywords. Furthermore, we will probably want to make statements about the functions of proteins more than about individual domains, so we should analyze the entities that we will later want to make inference about.

Overall strategy:
1. Extract the UniProt IDs of all human proteins that contain a PDZ domain from Pfam.
2. Extract all proteins with their keywords from UniProt (for simplicity, focusing only on human proteins).
3. Run an independent Fisher's exact test for each keyword, and finally correct for multiple testing (see details in the comments below).

Note that a chi-squared test cannot be used here for two reasons:
1. Keywords are NOT mutually exclusive.
2. We want to know WHICH of the keywords are significant (not only that there are indeed some significant keywords).

In [2]:
# From: ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam31.0/proteomes/9606.tsv.gz
human_pfam_records = pd.read_csv(os.path.join(DIR, '9606.tsv.gz'), sep = r'\t|> <', skiprows = 2, na_values = ['No_clan'], \
        engine = 'python').rename(columns = lambda name: re.sub(r'[ \-]', '_', re.sub(r'[#\<\>]', '', name)).lower())
human_pfam_domains = human_pfam_records[human_pfam_records['type'] == 'Domain']
display(human_pfam_records)

seq_id  alignment_start  alignment_end  envelope_start  \
0      A0A024QZ18               69            147              66   
1      A0A024QZ33                5            123               4   
2      A0A024QZ42               25             84              22   
3      A0A024QZB8               40            436              39   
4      A0A024QZP7                4            287               4   
5      A0A024QZX5               11            380              10   
6      A0A024R0K5               40            140              38   
7      A0A024R0K5              239            318             239   
8      A0A024R0K5              418            495             417   
9      A0A024R0K5              596            663             595   
10     A0A024R0K5              146            219             146   
11     A0A024R0K5              502            575             502   
12     A0A024R0K5              324            397             324   
13     A0A024R0T9               24            100              24   
14     A0A024R0Y4               74            117              72   
15     A0A024R0Y4              374            440             371   
16     A0A024R161               49            115              49   
17     A0A024R161              108            152              96   
18     A0A024R214              562            619             562   
19     A0A024R214              370            470             370   
20     A0A024R214               61            367              61   
21     A0A024R216                9             90               9   
22     A0A024R2X5               66            143              66   
23     A0A024R2X5              335            532             333   
24     A0A024R2X5              183            319             183   
25     A0A024R341                2             87               1   
26     A0A024R3B9                3             94               1   
27     A0A024R3M2               56            163              56   
28     A0A024R3M2              336            483             335   
29     A0A024R3W6               28            139              28   
...           ...              ...            ...             ...   
98206      X6RKN2              119            188             116   
98207      X6RKN2              511            596             509   
98208      X6RKN2              434            505             421   
98209      X6RKN2             1057           1140            1057   
98210      X6RKN2              240            307             230   
98211      X6RKN2               10             92              10   
98212      X6RKN2              330            399             325   
98213      X6RKQ2                2             45               2   
98214      X6RKS3              134            217             134   
98215      X6RKS3                1             28               1   
98216      X6RKW4              219            311             216   
98217      X6RKY7               14            153              14   
98218      X6RL08               10             87               9   
98219      X6RL45                9            141               1   
98220      X6RLA8               48            199              48   
98221      X6RLJ0              116            219             116   
98222      X6RLJ0               62            108              59   
98223      X6RLL4               15            286              15   
98224      X6RLP6               29             95              29   
98225      X6RLR1                7            160               6   
98226      X6RLT1               21            589              14   
98227      X6RLU5              109            188              82   
98228      X6RLX0             1072           1111            1072   
98229      X6RLX0              457            833             453   
98230      X6RLX0              154            461             154   
98231      X6RLX0              838            986             833   
98232    

In [3]:
pdz_protein_ids = set(human_pfam_domains.loc[human_pfam_domains['hmm_name'] == 'PDZ', 'seq_id'])
print('There are %d proteins with a PDZ domain.' % len(pdz_protein_ids))

There are 381 proteins with a PDZ domain.


In [4]:
'''
We build a dataframe of all ~20K reviewed human proteins from UniProt, and all the keywords of each of these proteins.
Note that the 'keywords' column is of type set (i.e. each entry in this column is a set of strings).
'''

uniprot_ids = []
keywords = []

# From: http://www.uniprot.org/uniprot/?query=organism%3A%22Homo+sapiens+%5B9606%5D%22+AND+reviewed%3Ayes&sort=score
with gzip.open(os.path.join(DIR, 'uniprot_human_reviewed.xml.gz'), 'rt') as f:
    for record in SeqIO.parse(f, 'uniprot-xml'):
        uniprot_ids.append(record.id)
        keywords.append(set(record.annotations['keywords']))
        
protein_data = pd.DataFrame({'uniprot_id': uniprot_ids, 'keywords': keywords})
display(protein_data)

uniprot_id                                           keywords
0         P32929  {Cysteine biosynthesis, 3D-structure, Lyase, A...
1         O14646  {Nucleus, Transcription, Mental retardation, P...
2         Q8IWX8  {Phosphoprotein, Polymorphism, Reference prote...
3         Q99653  {3D-structure, Cell membrane, Phosphoprotein, ...
4         O94983  {Alternative splicing, Polymorphism, Reference...
5         P42695  {Mitosis, Phosphoprotein, Polymorphism, Refere...
6         Q8IYT2  {S-adenosyl-L-methionine, Polymorphism, mRNA c...
7         Q9NSA3  {3D-structure, Phosphoprotein, Reference prote...
8         Q96KP4  {3D-structure, Carboxypeptidase, Alternative s...
9         O95476  {Polymorphism, Transmembrane helix, Reference ...
10        Q969H4  {3D-structure, Phosphoprotein, Alternative spl...
11        P18545  {Vision, 3D-structure, Polymorphism, Sensory t...
12        Q6PJW8  {Cell membrane, Alternative splicing, Cytoplas...
13        P26992  {3D-structure, Cell membrane, Reference proteo...
14        Q9NXG0  {Phosphoprotein, Alternative splicing, Polymor...
15        Q01955  {Deafness, Cell adhesion, Disulfide bond, Base...
16        P06681  {Innate immunity, Immunity, Disulfide bond, Se...
17        P27658  {Basement membrane, Collagen, Reference proteo...
18        P02461  {Calcium, Disulfide bond, Collagen, Polymorphi...
19        P0C0L5  {Cell projection, Inflammatory response, Syste...
20        P20908  {Calcium, Disulfide bond, Collagen, Polymorphi...
21        P25940  {Collagen, Polymorphism, Reference proteome, H...
22        P02748  {Target cell membrane, Age-related macular deg...
23        P10643  {Cytolysis, Complement alternate pathway, Inna...
24        Q13057  {Phosphoprotein, Alternative splicing, Polymor...
25        P00414  {Polymorphism, Primary mitochondrial disease, ...
26        A4D0Y5            {Reference proteome, Complete proteome}
27        Q49A92  {Polymorphism, Alternative splicing, Reference...
28        Q9UFW8  {Phosphoprotein, Transcription regulation, Ref...
29        Q96K31  {Alternative splicing, Reference proteome, Com...
...          ...                                                ...
20382     Q96EZ4  {Proto-oncogene, Polymorphism, Reference prote...
20383     Q8TBZ2  {Phosphoprotein, Developmental protein, Polymo...
20384     O00159  {Cell membrane, Cell projection, Actin-binding...
20385     P12829  {Muscle protein, Polymorphism, Reference prote...
20386     Q15746  {Cell projection, Actin-binding, Acetylation, ...
20387     Q9NPC7  {Zinc, 3D-structure, Phosphoprotein, Alternati...
20388     Q9Y2G1  {Nucleus, Acetylation, Transcription, Polymorp...
20389     Q8NFW9  {Zinc, Phosphoprotein, Alternative splicing, C...
20390     Q5VU43  {Phosphoprotein, Alternative splicing, Polymor...
20391     O15530  {Cell membrane, Ubl conjugation, Nucleus, Acet...
20392     Q6A1A2  {Phosphoprotein, Reference proteome, Transfera...
20393     Q9NUG6  {Cytoplasm, Reference proteome, Chaperone, Com...
20394     Q29RF7  {Mitosis, Phosphoprotein, Alternative splicing...
20395     Q96JP2  {3D-structure, Polymorphism, Reference proteom...
20396     Q99836  {3D-structure, Phosphoprotein, Alternative spl...
20397     Q9NR12  {Zinc, 3D-structure, Developmental protein, Al...
20398     P20592  {Nuclear pore complex, Antiviral defense, 3D-s...
20399     P35749  {Actin-binding, Proto-oncogene, Motor protein,...
20400     Q9Y6X6  {Phosphoprotein, Alternative splicing, Actin-b...
20401     O43795  {Phosphoprotein, Alternative splicing, Actin-b...
20402     P11055  {Muscle protein, Actin-binding, Polymorphism, ...
20403     Q9UM54  {Cell membrane, Cell projection, Actin-binding...
20404     P09131  {Polymorphism, Alternative splicing, Reference...
20405     Q9Y2Z2  {FAD, Alternative splicing, Primary mitochondr...
20406     Q6IN85  {Phosphoprotein, Alternative splicing, Referen...
20407     P49286  {Cell membrane, Polymorphism, Transmembrane he...
20408     P84157  {Phosphoprotein, Alternative splicing, Trans

In [5]:
all_keywords = set.union(*protein_data['keywords'])
print('There are %d unique keywords.' % len(all_keywords))

protein_data['contains_pdz'] = protein_data['uniprot_id'].isin(pdz_protein_ids)
print('%d of the %d reviewed proteins contain a PDZ domain.' % (protein_data['contains_pdz'].sum(), len(protein_data)))

display(protein_data.head())

There are 723 unique keywords.
129 of the 20412 reviewed proteins contain a PDZ domain.


uniprot_id                                           keywords  contains_pdz
0     P32929  {Cysteine biosynthesis, 3D-structure, Lyase, A...         False
1     O14646  {Nucleus, Transcription, Mental retardation, P...         False
2     Q8IWX8  {Phosphoprotein, Polymorphism, Reference prote...         False
3     Q99653  {3D-structure, Cell membrane, Phosphoprotein, ...         False
4     O94983  {Alternative splicing, Polymorphism, Reference...         False

In [16]:
'''
We iterate over all the unique keywords, and for each one we independently build a 2X2 contingency table. This table counts
how many proteins there are (of the entire ~20K reviewed human proteins) that have or don't have this keyword and have or don't
have a PDZ domain. The null hypothesis is that the rows and columns of this table are independent (i.e. no correlation between
the presence of the keyword and the presence of PDZ domain). We run a Fisher's exact test to calculate p-values. We also
calculate Risk-Ratio (RR) to measure the effect size.
'''

test_keywords = []
test_RRs = []
test_pvals = []

for keyword in all_keywords:
    
    # A mask matching to the protein_data dataframe for the given keyword (i.e. the mask will have a True value only
    # for proteins that have this keyword, False otherwise).
    keyword_mask = protein_data['keywords'].apply(lambda protein_keywords: keyword in protein_keywords)
    
    contingency_table = pd.DataFrame(0, index = ['pdz', 'no_pdz'], columns = ['keyword', 'no_keyword'])
    
    for row_index in range(2):
        
        # A mask over protein_data matching the row criterion (pdz or no_pdz)
        row_mask = protein_data['contains_pdz'] if row_index == 0 else ~protein_data['contains_pdz']
    
        for column_index in range(2):
            # A mask over protein_data matching the column criterion (keyword or no_keyword)
            column_mask = keyword_mask if column_index == 0 else ~keyword_mask
            # (row_mask & column_mask) will give the cell's specific mask (over protein_data). Summing this mask will give the
            # number of proteins that match the overall criterion for this cell (pdz vs. no_pdz AND keyword vs. no_keyword)
            contingency_table.iloc[row_index, column_index] = (row_mask & column_mask).sum()
    
    # The "risk" of having the keyword for a protein containing a PDZ domain
    pdz_risk = contingency_table.loc['pdz', 'keyword'] / contingency_table.loc['pdz', :].sum()
    # The "risk" of having the keyword for a protein NOT containing a PDZ domain
    no_pdz_risk = contingency_table.loc['no_pdz', 'keyword'] / contingency_table.loc['no_pdz', :].sum()
    
    if no_pdz_risk == 0:
        RR = np.nan
    else:
        RR = pdz_risk / no_pdz_risk
    
    _, pval = fisher_exact(contingency_table)
    
    test_keywords.append(keyword)
    test_RRs.append(RR)
    test_pvals.append(pval)
    
keyword_tests = pd.DataFrame({'keyword': test_keywords, 'RR': test_RRs, 'pval': test_pvals})
display(keyword_tests)

keyword        RR      pval
0                        Peroxisome biogenesis  0.000000  1.000000
1                                       Vision  1.379233  0.518638
2                                 Endonuclease  0.000000  1.000000
3                            Antiviral defense  0.000000  1.000000
4                         Sensory transduction  0.261618  0.190308
5                              Ubl conjugation  1.079109  0.686731
6                 Complement alternate pathway  0.000000  1.000000
7                              Transit peptide  0.000000  0.052706
8                              Target membrane  0.000000  1.000000
9                                   Paired box  0.000000  1.000000
10                      Cholesterol metabolism  0.000000  1.000000
11                                 Antioxidant  0.000000  1.000000
12                          Pheromone response  0.000000  1.000000
13                            DNA condensation  0.000000  1.000000
14                     Mast cell degranulation  0.000000  1.000000
15           Inhibition of host STAT1 by virus       NaN  0.006320
16                        Hirschsprung disease  0.000000  1.000000
17                   Phospholipid biosynthesis  0.000000  1.000000
18                              Opioid peptide  0.000000  1.000000
19                    Phospholipid degradation  0.000000  1.000000
20                      Multifunctional enzyme  0.000000  1.000000
21                                       Porin  0.000000  1.000000
22                          Electron transport  0.000000  1.000000
23                      Hydrogen ion transport  0.000000  1.000000
24                          Biological rhythms  0.000000  1.000000
25                         Translational shunt  0.000000  1.000000
26                               Mitochondrion  0.134732  0.007193
27                                    Dwarfism  2.041981  0.259002
28                                   Nitration  0.000000  1.000000
29                                      Virion  0.000000  1.000000
..                                         ...       ...       ...
693                                    Cadmium  0.000000  1.000000
694                          Glutathionylation  0.000000  1.000000
695                   Menaquinone biosynthesis  0.000000  1.000000
696                                Spliceosome  0.000000  1.000000
697                      Developmental protein  1.947152  0.034451
698                                GTP-binding  0.000000  0.285577
699                                     3Fe-4S  0.000000  1.000000
700                                   Protease  0.291712  0.268641
701                                    Hearing  7.669881  0.030231
702                             Fanconi anemia  0.000000  1.000000
703                DNA-directed RNA polymerase  0.000000  1.000000
704                                 Peroxidase  0.000000  1.000000
705                                  Microsome  0.000000  1.000000
706      Congenital stationary night blindness  0.000000  1.000000
707                                      Lyase  0.000000  1.000000
708                                   Oncogene  0.000000  1.000000
709  Hereditary nonpolyposis colorectal cancer  0.000000  1.000000
710                                  Glycation  0.000000  1.000000
711                                   Glaucoma  0.000000  1.000000
712                   Catecholamine metabolism  0.000000  1.000000
713                                   Telomere  0.000000  1.000000
714                                     Myosin  3.208828  0.271943
715                                    Mitosis  0.000000  0.424389
716              Amyotrophic lateral sclerosis  0.000000  1.000000
717                      Inositol biosynthesis  0.000000  1.000000
718                               Inflammasome  0.000000  1.000000
719                    Membrane attack complex  0.000000  1.000000
720                    Immunoglobulin V region  0.000000  1.000000
721                

In [17]:
EPSILON = 1e-5

# We must account for multiple testing!
keyword_tests['significance'], keyword_tests['qval'], _, _ = multipletests(keyword_tests['pval'], method = 'bonferroni')
significant_keyword_tests = keyword_tests[keyword_tests['significance']]
print('%d of %d keywords are significantly associated (or anti-associated) with the presence PDZ:' % \
        (len(significant_keyword_tests), len(keyword_tests)))

# We sort the significant keyword tests by max{RR, 1 / (RR + EPSILON)} (we use EPSILON to avoid division by zero)
significant_keyword_tests = significant_keyword_tests.assign(absolute_RR = significant_keyword_tests['RR'].apply(\
        lambda RR: max(RR, 1 / (RR + EPSILON)))).sort_values('absolute_RR', ascending = False).drop('absolute_RR', axis = 1)
display(significant_keyword_tests)

27 of 723 keywords are significantly associated (or anti-associated) with the presence PDZ:


keyword         RR          pval  significance  \
411                    Receptor   0.000000  4.593681e-05          True   
408                 DNA-binding   0.000000  2.159798e-06          True   
276              Usher syndrome  52.410853  5.203706e-05          True   
639              Tight junction  38.300239  1.911723e-23          True   
31                 Glycoprotein   0.033376  1.344335e-13          True   
552                  LIM domain  25.775829  2.377872e-11          True   
182                 Synaptosome  25.728964  2.491073e-10          True   
106              Disulfide bond   0.042052  2.562663e-10          True   
212                      Signal   0.042983  4.186233e-10          True   
591               Transmembrane   0.059955  4.303990e-14          True   
405         Transmembrane helix   0.060127  4.235922e-14          True   
340  Postsynaptic cell membrane  15.298303  7.090615e-16          True   
642    Transcription regulation   0.067165  4.807308e-06          True   
503                  SH3 domain  14.502032  2.614063e-16          True   
535               Cell junction  12.165439  2.897573e-42          True   
72                      Synapse  10.843625  4.175120e-19          True   
551               Actin-binding   7.801615  2.125775e-08          True   
335               Transcription   0.130700  2.638285e-05          True   
367             Cell projection   4.446622  4.477481e-10          True   
114                Cytoskeleton   3.943960  1.010603e-10          True   
57                Cell membrane   2.814905  3.215476e-16          True   
523                   Cytoplasm   2.700619  1.713493e-23          True   
270                3D-structure   2.154541  3.410318e-17          True   
124                      Repeat   2.130872  3.374656e-11          True   
668              Phosphoprotein   1.996419  4.051527e-20          True   
545                    Membrane   1.646414  6.266272e-08          True   
394        Alternative splicing   1.538833  1.218691e-10          True   

             qval  
411  3.321231e-02  
408  1.561534e-03  
276  3.762279e-02  
639  1.382175e-20  
31   9.719542e-11  
552  1.719202e-08  
182  1.801045e-07  
106  1.852805e-07  
212  3.026646e-07  
591  3.111785e-11  
405  3.062572e-11  
340  5.126515e-13  
642  3.475684e-03  
503  1.889967e-13  
535  2.094945e-39  
72   3.018612e-16  
551  1.536935e-05  
335  1.907480e-02  
367  3.237219e-07  
114  7.306661e-08  
57   2.324789e-13  
523  1.238855e-20  
270  2.465660e-14  
124  2.439876e-08  
668  2.929254e-17  
545  4.530515e-05  
394  8.811134e-08

Note that some of the significant keywords are associated with PDZ (RR > 1), while others are anti-associated (RR < 1)!

Note that the statistical test we carried out could be biased in many ways:

1. We rely on UniProt's annotations for keywords, which could be incomplete (lacking keywords that should have been part of a protein's annotation but weren't). Note that keywords and GO annotations are very sensitive to definitions and methodology, which vary between different databases. Different (yet equally justified) keyword definitions could have produced very different results.

2. We rely on Pfam's HMM model for defining the PDZ domain. Like with UniProt's keywords, there could be false positives and false negatives.

3. The existence of paralogs (duplications within the human proteome) could call to question our assumption that protein annotations are independent (Fisher's exact, like all tests, assumes i.i.d records).

4. We have looked only at human proteins, which could be a bad representation of the entire space of proteins. However, if we looked at all UniProt proteins, we would be susceptible to even greater biases due to:

    4.1 Overlapping between orthologs (homologs across different species) would probably dominate the analysis (much more than homologs within the same species) if we didn't account for them (e.g. by taking only "unique representatives" - this is quite challenging to define).
    
    4.2. The sapce of species covered in UniProt is biased by the species that have been more studied (like all databases that attempt to cover all organisms).
    
    4.3. Even when a species is represented in UniProt, the quality of annotations (e.g. of the keywords) is not identical across the board. More studied specieis should have better, more accurate annotations, and the annotations for human proteins are probably the most comprehensive. The same goes for Pfam annotations.

    For all of these reasons, the analysis was carried out only for human proteins. But it means that the results should not be blindly generalized to other species.
    

Other than these potential biases, it is expected that the associations between various functional proteomic keywords to the presence of PDZ domains are due to genuine functional characteristics of PDZ-containing proteins. Such statistical characteristics can be directly dervied from the function of PDZ, or be indirectly associated. For example, the keyword "Postsynaptic cell membrane" is directly related to the role of PDZ in the brain, so it's not surprising to find it very significantly and strongly enriched in PDZ-containing proteins. The depletion of the "DNA-binding" keyword in PDZ-containing proteins, on the other hand, is more indirect, but not surprising: membrane proteins are less likely to find their way into the nucleus and bind DNA.

Enrichment of the following keywords is expected given the known roles of PDZ: Synaptosome, Postsynaptic cell membrane, Cell junction, Synapse, Cytoskeleton, Cell membrane, Phosphoprotein, Membrane.

Depletion of the following keywords is more surprising and unexpected: Glycoprotein, Signal, Transmembrane. 